manage_breast_screening/clinics/models.py

```diff
@@ -189,6 +192,24 @@ class UserAssignment(BaseModel):
     provider = models.ForeignKey(
         Provider, on_delete=models.PROTECT, related_name="assignments"
     )
+    roles = ArrayField(
+        base_field=models.CharField(
+            max_length=32,
+            choices=[
+                (Role.CLINICAL.value, Role.CLINICAL.value),
+                (Role.ADMINISTRATIVE.value, Role.ADMINISTRATIVE.value),
+            ],
+        ),
+        default=list,
+        validators=[MinLengthValidator(1)],
+        help_text="Roles granted to the user for this provider.",
+    )
+
+    def save(self, *args, **kwargs):
+        if self.roles:
+            # Remove duplicates and sort
```

Comment:
this doesn't completely guarantee that `roles` has only unique values, since you can create or update via queryset methods and bypass model instances. Alternatively you could move this bit up into the authorization layer.